### ISSUE TITLE
Strange issue with C++ 20 feature(s)

### ISSUE THREAD

**Issue by cpsauer** [clangd/clangd#1079]
Hi awesome clangd folks,

Someone using the [Bazel adapter I'd built ran into a strange issue with clangd and C++20](https://github.com/hedronvision/bazel-compile-commands-extractor/issues/32).

I've cooked up a super minimal example based on theirs. We've got a simple `main` that uses a C++20 feature, `std::construct_at`. It looks like this:

```c++
#include <memory>

int main() {
    int i = 0;
    std::construct_at(&i, 0);
    return EXIT_SUCCESS;
}

```

It builds (and runs) fine with `clang main.cpp -std=c++20`. However, specifying that command in `compile_commands.json` leads to a red squiggly error: `No member named 'construct_at' in namespace 'std' clang(no_member)`.

Curiously, changing the command in `compile_commands.json` to use an incorrect executable,  say `chmod` as an example of something obviously not a compiler, leads to things working.

Here's the `compile_comman.json` file:
```
[{
    "file": "main.cpp",
    "command": "chmod -std=c++20",
    "directory": "/Users/cpsauer/Downloads/clangdIssue"
}]
```

And here's the whole thing zipped up for easy experimenting (though you'll need to change the directory in `compile_comman.json`): 
[clangdIssue.zip](https://github.com/clangd/clangd/files/8347516/clangdIssue.zip)

Any idea what's going on?

Thanks for reading and for all you do!
Chris

P.S. Adding --query-driver doesn't seem to help.

System information
Output of clangd --version: 13.0.0
Editor/LSP plugin: VSCode
Operating system: macOS 12

**Comment by HighCommander4** [comment]
Could you try the following please:

1. In the working case (`chmod` in the command), follow the `<memory>` include, and take note of the full path of the opened file.
2. In the broken case, do the same thing. Is it the same path or different?

My hypothesis is that you may be getting different standard library installations in the two cases, though the mechanism by which that happens is not clear yet.

If you're using `--query-driver`, please mention its value as well (or better, the full clangd command line).

**Comment by cpsauer** [comment]
Hey, @HighCommander4! Of course.

The paths do indeed differ. Good call.

For `chmod` it's `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/memory`
For `clang` it's 
`/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory`

Setting some common SDK flags (as below) doesn't change this:
```
-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -F/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks -F/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks
```

Re `--query-driver`: all of the above happened originally without. But after further fiddling, I can get things to work properly by whitelisting everything. That is, the (only) flag specified for clangd is `--query-driver=/**/*` This also changes the path above to match that for `chmod` that works.

In the clangd output it says:

```
System includes extractor: successfully executed /usr/bin/clang
	got includes: "/usr/local/include, /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1, /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/13.1.6/include, /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include, /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include, /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)"
```

That's definitely a good workaround, but I won't close because think clangd aims to auto-discover on macOS (per https://clangd.llvm.org/guides/system-headers#how-clangd-finds-those-headers), right?

Thanks so much for all your good help, Nathan :) I really appreciate you.

Chris

**Comment by HighCommander4** [comment]
Very curious. I don't know enough about macOS to say how those two standard library installations might differ, but I'm guessing the "bad" one (the one in `Toolchains/XcodeDefault.xctoolchain`) might be an older version that predates the addition of `std::construct_at` in the C++20 timeframe.

By the way, I learned something about `--query-driver` today: if the compiler in the compile command is a relative path (e.g. `clang`), it's turned into an absolute path first (e.g. `/usr/bin/clang`), and _that's_ what's checked against the `--query-driver` regex. So, if you find using a less permissive regex to be a slightly more palatable workaround, I think `--query-driver=/usr/bin/clang` should work.

**Comment by cpsauer** [comment]
Yeah, that seems to be what's going on. <memory> still seems get a C++ 20 version of __cplusplus (202002L) and therefore gets a #define _LIBCPP_STD_VER 18 in __config, just like the other file, but I don's see a __memory subfolder or a construct_at.h 

I mean, honestly, I'm pretty down to just have --query-driver on by default and to recommend that to our (mutual) users. Solves some other issues with Bazel wrapping compilers in shell scripts.

But I think in general clangd wants to work without --query-driver for clang on macOS, right?

**Comment by HighCommander4** [comment]
My mental model so far was that, without `--query-driver`, clangd uses some default set of system include paths, but having read the [guide](https://clangd.llvm.org/guides/system-headers#how-clangd-finds-those-headers) you linked to above (thanks for sharing!), I now understand that clangd actually uses heuristics that still take into account the compiler from the compile command (just without executing it to query its include paths).

So I think that explains on a high level why having `clang` vs. `chmod` there makes a difference.

Assuming that `chmod` doesn't lead to any useful heuristics and causes clangd to use some fallback set of system include paths, which in this case are correct -- the fact that having the correct compiler (`clang`) there makes the outcome of the heuristics **worse**, definitely seems like a bug to me.

**Comment by cpsauer** [comment]
#487 might be related!

**Comment by mateusz-plociennik** [comment]
Have you tried providing full `Xcode` path to `--query-driver=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/*` parameter? 

After I provided such parameter, the `clangd` stopped complaining about `no_member` stuff.

In my `CMake` generated `compile_commands.json` the `command` starts with `/Applications/Xcode-15.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++`.

**Comment by cpsauer** [comment]
Hi Matheuz! Thanks for engaging. --query-driver works (as above)--but this issue is about problems in the auto-discovery on macOS.
